internal/provider/resource_secret.go

```diff
@@ -0,0 +1,299 @@
+	})
+	if err != nil {
+		resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to add secret, got error: %s", err))
+		return
+	}
+	s.trace(fmt.Sprintf("add secret resource %q", plan.SecretId))
+
+	plan.SecretId = types.StringValue(createSecretOutput.SecretId)
+
+	// Save plan into Terraform state
+	resp.Diagnostics.Append(resp.State.Set(ctx, &plan)...)
+
+	s.trace(fmt.Sprintf("create secret resource %q", plan.SecretId))
+}
+
+func (s *secretResource) Read(ctx context.Context, req resource.ReadRequest, resp *resource.ReadResponse) {
+	// Prevent panic if the provider has not been configured.
+	if s.client == nil {
+		addClientNotConfiguredError(&resp.Diagnostics, "secret", "read")
+		return
+	}
+
+	var state secretResourceModel
+
+	// Read Terraform configuration state into the model
+	resp.Diagnostics.Append(req.State.Get(ctx, &state)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	readSecretOutput, err := s.client.Secrets.ReadSecret(&juju.ReadSecretInput{
+		SecretId:  state.SecretId.ValueString(),
+		ModelName: state.Model.ValueString(),
+		Name:      state.Name.ValueStringPointer(),
+		Revision:  nil,
+	})
+	if err != nil {
+		resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to read secret, got error: %s", err))
+		return
+	}
+	s.trace(fmt.Sprintf("read secret resource %q", state.SecretId))
+
+	// Print the secret details
+	fmt.Printf("r = %#v\n", readSecretOutput)
+
+	// Save the secret details into the Terraform state
+	if !state.Name.IsNull() {
+		state.Name = types.StringValue(readSecretOutput.Name)
+	}
+	if !state.Info.IsNull() {
+		state.Info = types.StringValue(readSecretOutput.Info)
+	}
+
+	secretValue, errDiag := types.MapValueFrom(ctx, types.StringType, readSecretOutput.Value)
+	resp.Diagnostics.Append(errDiag...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+	state.Value = secretValue
+
+	// Save state into Terraform state
+	resp.Diagnostics.Append(resp.State.Set(ctx, &state)...)
+}
+
+func (s *secretResource) Update(ctx context.Context, req resource.UpdateRequest, resp *resource.UpdateResponse) {
+	// Prevent panic if the provider has not been configured.
+	if s.client == nil {
+		addClientNotConfiguredError(&resp.Diagnostics, "secret", "update")
+		return
+	}
+
+	var plan, state secretResourceModel
+
+	// Read Terraform plan and state into the model
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &plan)...)
+	resp.Diagnostics.Append(req.State.Get(ctx, &state)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// print the plan and state
+	fmt.Printf("plan = %+v\n", plan)
+	fmt.Printf("state = %+v\n", state)
+
+	var err error
+	noChange := true
+
+	var updatedSecretInput juju.UpdateSecretInput
+
+	updatedSecretInput.ModelName = state.Model.ValueString()
+	updatedSecretInput.SecretId = state.SecretId.ValueString()
+
+	// Check if the secret name has changed
+	if !plan.Name.Equal(state.Name) {
+		noChange = false
+		updatedSecretInput.Name = plan.Name.ValueStringPointer()
```

Comment:
You shouldn't do this if there is no change. Only set updatedSecretInput.Name if there IS a change.

Reply:
fixed